Remove drop order twist of && and || and make them associative

est31 opened the pull request:
Previously a short circuiting binop chain (chain of && or ||s) would drop the temporaries created by the first element after all the other elements, and otherwise follow evaluation order. So `f(1).g() && f(2).g() && f(3).g() && f(4).g()` would drop the temporaries in the order `2,3,4,1`. This made `&&` and `||` non-associative regarding drop order. In other words, adding ()'s to the expression would change drop order: `f(1).g() && (f(2).g() && f(3).g()) && f(4).g()` for example would drop in the order `3,2,4,1`.

As, except for the bool result, there is no data returned by the sub-expressions of the short circuiting binops, we can safely discard of any temporaries created by the sub-expr. Previously, code was already putting the rhs's into terminating scopes, but missed it for the lhs's.

This commit addresses this "twist". We now also put the lhs into a terminating scope. The drop order of the above expressions becomes `1,2,3,4`.

There might be code relying on the current order, and therefore I'd recommend doing a crater run to gauge the impact. I'd argue that such code is already quite wonky as it is one `foo() &&` addition away from breaking. ~~For the impact, I don't expect any *build* failures, as the compiler gets strictly more tolerant: shortening the lifetime of temporaries only expands the list of programs the compiler accepts as valid. There might be *runtime* failures caused by this change however.~~ Edit: both build and runtime failures are possible, e.g. see the example provided by dtolnay [below](https://github.com/rust-lang/rust/pull/103293#issuecomment-1285341113). Edit2: the crater run has finished and [results](https://github.com/rust-lang/rust/pull/103293#issuecomment-1292275203) are that there is only one build failure which is easy to fix with a +/- 1 line diff.

I've included a testcase that now compiles thanks to this patch.

The breakage is also limited to drop order relative to conditionals in the && chain: that is, in code like this:

```Rust
let hello = foo().hi() && bar().world();
println!("hi");
```

we already drop the temporaries of `foo().hi()` before we reach "hi".

I'd ideally have this PR merged before let chains are stabilized. If this PR is taking too long, I'd love to have a more restricted version of this change limited to `&&`'s in let chains: the `&&`'s of such chains are quite special anyways as they accept `let` bindings, in there the `&&` is therefore more a part of the "if let chain" construct than a construct of its own.

Fixes #103107

Status: waiting on [this accepted FCP](https://github.com/rust-lang/rust/pull/103293#issuecomment-1293411354) finishing.

est31 committed:
Remove drop order twist of && and || and make them associative

Previously a short circuiting && chain would drop the
first element after all the other elements, and otherwise
follow evaluation order, so code like:

f(1).g() && f(2).g() && f(3).g() && f(4).g()

would drop the temporaries in the order 2,3,4,1. This made
&& and || non-associative regarding drop order, so
adding ()'s to the expression would change drop order:

f(1).g() && (f(2).g() && f(3).g()) && f(4).g()

for example would drop in the order 3,2,4,1.

As, except for the bool result, there is no data returned
by the sub-expressions of the short circuiting binops,
we can safely discard of any temporaries created by the
sub-expr. Previously, code was already putting the rhs's
into terminating scopes, but missed it for the lhs's.

This commit addresses this "twist". In the expression,
we now also put the lhs into a terminating scope.
The drop order for the above expressions is 1,2,3,4
now.

rust-highfive commented:
r? @wesleywiser

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3288174221/jobs/5418194179) [(plain)](https://github.com/rust-lang/rust/commit/6172afd6ed77f5a83fc8a2f05ed8246277ff598f/checks/9002723800/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_hir_analysis/src/check/region.rs at line 259:
                 // being dropped last. For the Let exception, see below.
                 let terminate_lhs = match l.kind {
                     hir::ExprKind::Let(_) => false,
-                    hir::ExprKind::Binary(source_map::Spanned { node, .. }, ..) if node == outer => false,
+                    hir::ExprKind::Binary(source_map::Spanned { node, .. }, ..)
+                        if node == outer =>
+                        false
+                    }
                     _ => true,
                 };
                 };
                 if terminate_lhs {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_hir_analysis/src/check/dropck.rs" "/checkout/compiler/rustc_hir_analysis/src/check/intrinsicck.rs" "/checkout/compiler/rustc_hir_analysis/src/check/upvar.rs" "/checkout/compiler/rustc_hir_analysis/src/check/region.rs" "/checkout/compiler/rustc_hir_analysis/src/check/pat.rs" "/checkout/compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs" "/checkout/compiler/rustc_hir_analysis/src/check/_match.rs" "/checkout/compiler/rustc_hir_analysis/src/check/expectation.rs"` failed.
Build completed unsuccessfully in 0:00:12
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

dtolnay commented:
> For the impact, I don't expect any _build_ failures, as the compiler gets strictly more tolerant: shortening the lifetime of temporaries only expands the list of programs the compiler accepts as valid.

This is not quite right, for example this PR would break the following program.

```rust
struct Thing;

impl Drop for Thing {
    fn drop(&mut self) {}
}

struct Ref<'a>(&'a Thing);

impl<'a> Drop for Ref<'a> {
    fn drop(&mut self) {}
}

fn main() {
    let tmp;
    let _ = match Thing {
        ref thing => {
            tmp = thing;
            true
        }
    } && match Ref(tmp) {
        _ => false,
    };
}
```

It seems weird but it's not that much weirder than things like https://github.com/dtolnay/syn/blob/1.0.102/src/data.rs#L289-L293, so it's hard for me to gauge how widespread this might be.

est31 commented:
Good point, I confirm this fails with the PR applied. Text edited.

joshtriplett commented:
I think this is worth a crater run, and worth attempting even if we need to fix a couple of crates; this is a substantial improvement to understandability.

With this change, will that also make let chains able to use a consistent order?

est31 commented:
@joshtriplett indeed my investigation of how let chains work has motivated this change. `let chains` are affected by this twist, yes: if there is a bool condition at the very start of a let chain, it will be dropped last (but if there is any number of let conditions before, this twist doesn't activate, so it's not just "first bool condition comes last"). After this PR, the drop order rule for let chains will be simple: first bool conditions are dropped in reading order, then comes the body, then let conditions in reverse order (as the body might rely on them, as well as the let conditions on each other, so reverse order is the only one you can really do here).

The other behaviour that from my POV should be improved is the drop order of let temporaries regarding the else block. You can either choose consistency with `if let` and `match` here, or on the other side, with `if` and `let else`. Personally I'd prefer the latter two as it frees up resources earlier, but feel free to disagree (it might be surprising to people that adding/removing a bool conditional changes what the still-existing `let` bindings are doing regarding the `else` drop order). The entire topic of what a good drop order would be would benefit from lang team input. I've made a longer description of behaviour and how I would like it to change [here](https://github.com/rust-lang/rust/issues/100513#issuecomment-1279908873). I've edited the comment to reflect recent changes so it should be up to date.

est31 commented:
@bors try

bors commented:
:hourglass: Trying commit 0e70267b50790094287a0d6711fab2849b9fe948 with merge 51ad93974f526eaf4ede64932870490e6c96e4fd...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0e70267b50790094287a0d6711fab2849b9fe948","merge_sha":"51ad93974f526eaf4ede64932870490e6c96e4fd"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3292127763/jobs/5428620845)
Build commit: 51ad93974f526eaf4ede64932870490e6c96e4fd (`51ad93974f526eaf4ede64932870490e6c96e4fd`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3292127763/jobs/5428620845"},"merge_sha":"51ad93974f526eaf4ede64932870490e6c96e4fd"} -->

Dylan-DPC commented:
@craterbot check

est31 commented:
So it seems that there is one build failure due to the drop order changes. It occurs in a match guard in `spaad_internal-0.4.0`.

Relevant [log](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/reg/spaad_internal-0.4.0/log.txt) excerpt:

```
[INFO] [stdout] error[E0716]: temporary value dropped while borrowed
[INFO] [stdout]    --> src/entangle/transform.rs:459:51
[INFO] [stdout]     |
[INFO] [stdout] 459 |         ReturnType::Type(_, ty) if !(is_name(ty, &name.to_string()) || is_name(ty, "Self"))
[INFO] [stdout]     |                                                   ^^^^^^^^^^^^^^^^-    ------- borrow later used here
[INFO] [stdout]     |                                                   |               |
[INFO] [stdout]     |                                                   |               temporary value is freed at the end of this statement
[INFO] [stdout]     |                                                   creates a temporary which is freed while still in use
[INFO] [stdout]     |
[INFO] [stdout] help: consider using a `let` binding to create a longer lived value
[INFO] [stdout]     |
[INFO] [stdout] 457 ~     let binding = name.to_string();
[INFO] [stdout] 458 ~     if matches!(
[INFO] [stdout] 459 |         &sig.output,
[INFO] [stdout] 460 ~         ReturnType::Type(_, ty) if !(is_name(ty, &binding) || is_name(ty, "Self"))
[INFO] [stdout]     |
```

[This](https://github.com/Restioson/spaad/blob/7c3f7c43efc9e36384a4c85fe4d0ad6bc1737c2f/spaad_internal/src/entangle/transform.rs#L459) is the relevant line on github.

I don't really understand how this change can cause the build failure, as name is a const reference. All the other involved variables are references too. I don't get why the borrow should later be used. It seems like a bug in rustc to me, that I've missed something when writing my PR. It might be interesting to create a minimum reproducing example.

For the other failures, most of them are either in proc macros or build.rs failures. I need to investigate more closely, but a *lot* of crate failures are due to libc's build.rs failing:

```
[INFO] [stderr] error: failed to run custom build command for `libc v0.2.93`
[INFO] [stderr] 
[INFO] [stderr] Caused by:
[INFO] [stderr]   process didn't exit successfully: `/opt/rustwide/target/debug/build/libc-249f7f643d38f60c/build-script-build` (exit status: 101)
[INFO] [stderr]   --- stdout
[INFO] [stderr]   cargo:rerun-if-changed=build.rs
[INFO] [stderr] 
[INFO] [stderr]   --- stderr
[INFO] [stderr]   thread 'main' panicked at 'Failed to get rustc version', /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/libc-0.2.93/build.rs:9:63
```

The [code in libc](https://github.com/rust-lang/libc/blob/71b864d3772fce9e1439556b0e0fa2af6bbeaa14/build.rs#L115-L148) doesn't use any && or || except for one place where it uses ||.

I'm not sure what is going on here. I think though that many more crates depend on `libc` than the 44 failed crates detected by crater, so it can't be a general build failure of libc.

est31 commented:
So a minimum reproducing example is 
```Rust
fn main() {
    fn is_empty(s: &str) -> bool {
        s.is_empty()
    }
    // Doesn't fail if provided with this
    //let check_val: fn(&str) -> bool = |v| is_empty(v);
    let check_val = |v| is_empty(v);
    let n = "Hi";
    if check_val(&n.to_string()) || check_val("Hello") {}
}
```

```
error[E0716]: temporary value dropped while borrowed
  --> file.rs:10:19
   |
LL |     if check_val(&n.to_string()) || check_val("Hello") {}
   |                   ^^^^^^^^^^^^^-    --------- borrow later used here
   |                   |            |
   |                   |            temporary value is freed at the end of this statement
   |                   creates a temporary which is freed while still in use
   |
   = note: consider using a `let` binding to create a longer lived value
```

And yes, on current Rust it fails if you add a `false ||` to this chain, or move the two checks out like:

```Rust
check_val(&n.to_string());
check_val("Hello");
```

There seems to be something going on in the auto-generation of `check_val`'s signature that makes it fail: if you provide the signature for check_val yourself then it won't issue the error. It will issue the error if you specify it as `fn(_) -> bool` but it won't issue it if you specify it as `fn(&_) -> bool`. Something really interesting is going on here, maybe even a bug, but it's not a bug of my PR.

The other errors I still need to investigate.

est31 commented:
## 2 times panic of vulkano proc-macro derive:

Example [log](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/gh/gclark916.vulkano_examples/log.txt) excerpt:

```
[INFO] [stdout] error: proc-macro derive panicked
[INFO] [stdout]   --> src/main.rs:83:18
[INFO] [stdout]    |
[INFO] [stdout] 83 |         #[derive(VulkanoShader)]
[INFO] [stdout]    |                  ^^^^^^^^^^^^^
[INFO] [stdout]    |
[INFO] [stdout]    = help: message: Failed to execute glslangValidator: Os { code: 8, kind: Uncategorized, message: "Exec format error" }
```

I couldn't reproduce it when cloning https://github.com/gclark916/vulkano_examples . It seems to work for me. I think the bug is spurious.

## 1 time spaad_internal failure

Example [log](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/reg/spaad_internal-0.4.0/log.txt) excerpt:

```
[INFO] [stdout] error[E0716]: temporary value dropped while borrowed
[INFO] [stdout]    --> src/entangle/transform.rs:459:51
[INFO] [stdout]     |
[INFO] [stdout] 459 |         ReturnType::Type(_, ty) if !(is_name(ty, &name.to_string()) || is_name(ty, "Self"))
[INFO] [stdout]     |                                                   ^^^^^^^^^^^^^^^^-    ------- borrow later used here
[INFO] [stdout]     |                                                   |               |
[INFO] [stdout]     |                                                   |               temporary value is freed at the end of this statement
[INFO] [stdout]     |                                                   creates a temporary which is freed while still in use
```

I have written about this [above](https://github.com/rust-lang/rust/pull/103293#issuecomment-1288197313), but it is a confirmed regression of this PR. It is an easy build failure to fix however, and a PR to that repo should be easy.

## 2 file already exists error

Example [log](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/gh/Ayrx.binja-rs-hello-world/log.txt) excerpt: 

```
[INFO] [stderr] error: failed to run custom build command for `binaryninjacore-sys v0.1.0 (https://github.com/Vector35/binaryninja-api.git?branch=dev#661d7c95)`
[INFO] [stderr] 
[INFO] [stderr] Caused by:
[INFO] [stderr]   process didn't exit successfully: `/opt/rustwide/target/debug/build/binaryninjacore-sys-8a29f4355f98035e/build-script-build` (exit status: 101)
[INFO] [stderr]   --- stdout
[INFO] [stderr]   cargo:rerun-if-changed=../../binaryninjacore.h
[INFO] [stderr] 
[INFO] [stderr]   --- stderr
[INFO] [stderr]   thread 'main' panicked at 'failed to create required symlink: Os { code: 17, kind: AlreadyExists, message: "File exists" }', /opt/rustwide/cargo-home/git/checkouts/binaryninja-api-3c86176688a74ccb/661d7c9/rust/binaryninjacore-sys/build.rs:70:14
[INFO] [stderr]   stack backtrace:
```

I think this might be due to how rustwide is running things. Link to the [code](https://github.com/Vector35/binaryninja-api/blob/f2b60cdb7d9ce0494b0e2ff9e95c34bb3c487276/rust/binaryninjacore-sys/build.rs#L70):

```Rust
        let symlink_target = PathBuf::from(&out_dir).join("libbinaryninjacore.so");
        if !link_path.join("libbinaryninjacore.so").exists() && !symlink_target.exists() {
            use std::os::unix::fs;
            fs::symlink(
                link_path.join("libbinaryninjacore.so.1"),
                PathBuf::from(&out_dir).join("libbinaryninjacore.so"),
            )
            .expect("failed to create required symlink");
        }
```

It seems to erroneously compute the symlink target, and if I compare the master vs branch version of the build logs I see some downloading going on in the master version that my PR's build does not have. This makes me assume that the PR build shares data with the build from the master branch, in other words, the symlink is there from the master branch build putting it there.

## 27 libc failed to get rustc version

Example [log](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/gh/Xijezu.ServerMonitor/log.txt) excerpt:

```
[INFO] [stderr] error: failed to run custom build command for `libc v0.2.51`
[INFO] [stderr] 
[INFO] [stderr] Caused by:
[INFO] [stderr]   process didn't exit successfully: `/opt/rustwide/target/debug/build/libc-8fcc20d1842da443/build-script-build` (exit status: 101)
[INFO] [stderr]   --- stderr
[INFO] [stderr]   thread 'main' panicked at 'Failed to get rustc version', /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/libc-0.2.51/build.rs:7:31
```

I couldn't reproduce this locally when cloning https://github.com/letaron/rumake . I couldn't find any use of && or || in the relevant libc code, see [above](https://github.com/rust-lang/rust/pull/103293#issuecomment-1288197313). I think the failure is spurious, however that it occurs 27 times is a bit troubling. I would love to get input from people experienced with crater runs -- @Mark-Simulacrum are you familiar with this failure occuring?

## 3 error: expected expression, found `<eof>`

Example [log](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/gh/aki-akaguma.cmp_template_engine/log.txt) excerpt:

```
[INFO] [stdout] error: expected expression, found `<eof>`
[INFO] [stdout]  --> /opt/rustwide/target/debug/build/sailfish-compiler-d2f5ef8f8c6f2180/out/templates/big-table-6374291007e0e361:0:1
[INFO] [stdout]   |
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] error: non-statement macro in statement position: include
[INFO] [stdout]  --> src/ben_sailfish.rs:5:10
[INFO] [stdout]   |
[INFO] [stdout] 5 | #[derive(TemplateOnce)]
[INFO] [stdout]   |          ^^^^^^^^^^^^
[INFO] [stdout]   |
[INFO] [stdout]   = note: this error originates in the derive macro `TemplateOnce` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Other [example](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/gh/mrrbrilliant.learn-sailfish/log.txt).

I couldn't reproduce this locally when cloning https://github.com/aki-akaguma/cmp_template_engine . I think the failure is spurious.

## 1 build.rs failure for protobuf

Example [log](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/gh/ampretia.rust-datastructure-protobuf/log.txt):

```
[INFO] [stderr] error: failed to run custom build command for `protobuf v2.17.0`
[INFO] [stderr] 
[INFO] [stderr] Caused by:
[INFO] [stderr]   process didn't exit successfully: `/opt/rustwide/target/debug/build/protobuf-95e2dcd7a29b0510/build-script-build` (exit status: 101)
[INFO] [stderr]   --- stderr
[INFO] [stderr]   /opt/rustwide/rustup-home/toolchains/51ad93974f526eaf4ede64932870490e6c96e4fd/bin/rustc: symbol lookup error: /opt/rustwide/rustup-home/toolchains/51ad93974f526eaf4ede64932870490e6c96e4fd/lib/librustc_driver-9b1cd69a927aa865.so: undefined symbol: _ZN9hashbrown3raw11Fallibility17capacity_overflow17hf80e9b1ba4f59154E
[INFO] [stderr]   thread 'main' panicked at 'assertion failed: child.wait().expect(\"wait\").success()', /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/protobuf-2.17.0/build.rs:35:5
```

I have no idea what is going on here. It seems to be some kind of startup failure of rustc?

## 1 not a directory error

Example [log](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/gh/stuartZhang.scaffold-wizard/log.txt):

```
[INFO] [stderr] Caused by:
[INFO] [stderr]   process didn't exit successfully: `/opt/rustwide/target/debug/build/scaffold-wizard-0f71bb76ad80d23a/build-script-build` (exit status: 101)
[INFO] [stderr]   --- stdout
[INFO] [stderr]   调试：OUT_DIR=/opt/rustwide/target/debug/build/scaffold-wizard-95a67c924cea2612/out
[INFO] [stderr] 
[INFO] [stderr]   --- stderr
[INFO] [stderr]   thread 'main' panicked at '失败：不能删除原来的 /opt/rustwide/target/assets 符号链接文件: Os { code: 20, kind: NotADirectory, message: "Not a directory" }', build.rs:98:14
```

[The code](https://github.com/stuartZhang/scaffold-wizard/blob/afa3f1ff91917d552310e43422013da28e9f05c5/build.rs#L98):

```Rust
    if assets_symbol.exists() {
        fs::remove_dir(assets_symbol.clone())
            .expect(&format!("失败：不能删除原来的 {} 符号链接文件", assets_symbol.display())[..]);
    }
```

I don't know any chinese (i dont know even if it *is* chinese as many languages use the script), but it seems to fail to remove some directories. I think this is like the symlink issue above, an artifact of how rustwide runs tests.

## 4 no space left on device error

Example [log](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/gh/tmiskow.instant-compiler/log.txt):

```
[INFO] [stderr]    Compiling serde_derive v1.0.102
[INFO] [stderr] LLVM ERROR: IO failure on output stream: No space left on device
[INFO] [stderr] error: could not compile `serde_derive`
```

Other [example](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/gh/umstek.edge-comments/log.txt).

[Variation](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/gh/waycrate.wayout/log.txt):

```
[INFO] [stderr]   thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 28, kind: StorageFull, message: "No space left on device" }', /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/wayland-scanner-0.29.4/src/lib.rs:156:40
```

Other [variation](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/gh/wezm.steno-lookup/log.txt):

```
[INFO] [stderr] Error response from daemon: mkdir /var/lib/docker/overlay2/34cc4ed63b6f8398262113deb2cc78231ac5f41a2ccb4b24d3e3ed40b7730786/merged: no space left on device
```

I think that this error is spurious and was caused by the build environment.

## 2 ICE 'assertion failed: pos.get() <= self.position()'

Example [log](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/gh/zakaluka.egui_test_one/log.txt):
```
[INFO] [stderr] thread 'rustc' panicked at 'assertion failed: pos.get() <= self.position()', compiler/rustc_metadata/src/rmeta/encoder.rs:425:9
[INFO] [stderr] error: could not compile `wayland-protocols`
```

Other [example](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/gh/wfraser.unicoder/log.txt).

This might be related to the space problems? I've tried compiling wayland-protocols for some other projects with the local compiler, didn't notice any issues.

## 2 rustfmt not available

Example [log](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/reg/catboost2-0.1.1%2Bcatboost.1.0.5/log.txt):
```
[INFO] [stderr] error: failed to run custom build command for `catboost2-sys v0.1.1+catboost.1.0.5`
[INFO] [stderr] 
[INFO] [stderr] Caused by:
[INFO] [stderr]   process didn't exit successfully: `/opt/rustwide/target/debug/build/catboost2-sys-58e9375bdc591b38/build-script-build` (exit status: 101)
[INFO] [stderr]   --- stderr
[INFO] [stderr]   error: 'rustfmt' is not installed for the toolchain '51ad93974f526eaf4ede64932870490e6c96e4fd'
[INFO] [stderr]   note: this is a custom toolchain, which cannot use `rustup component add`
[INFO] [stderr]   help: if you built this toolchain from source, and used `rustup toolchain link`, then you may be able to build the component with `x.py`
[INFO] [stderr]   Failed to run rustfmt: Internal rustfmt error (non-fatal, continuing)
[INFO] [stderr]   thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 17, kind: AlreadyExists, message: "File exists" }', /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/catboost2-sys-0.1.1+catboost.1.0.5/build.rs:21:33
```

Other [example](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/reg/catboost2-sys-0.1.1%2Bcatboost.1.0.5/log.txt).

I think this is caused by how crater builds things instead of this PR's changes, and there is a PR to add the crates as flaky: https://github.com/rust-lang/crater/pull/667

## 2 proc-macro-error None unwrap

Example [log](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/reg/proc_unroll-0.1.1/log.txt):

```
[INFO] [stderr] error: failed to run custom build command for `proc-macro-error v1.0.4`
[INFO] [stderr] 
[INFO] [stderr] Caused by:
[INFO] [stderr]   process didn't exit successfully: `/opt/rustwide/target/debug/build/proc-macro-error-835051a2e0e7bdf1/build-script-build` (exit status: 101)
[INFO] [stderr]   --- stdout
[INFO] [stderr]   cargo:rustc-cfg=use_fallback
[INFO] [stderr] 
[INFO] [stderr]   --- stderr
[INFO] [stderr]   thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/proc-macro-error-1.0.4/build.rs:7:44
```

Other [example](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/reg/roopert_macro_root-0.2.0/log.txt).

[Link](https://gitlab.com/CreepySkeleton/proc-macro-error/-/blob/75f7d521387b58e20960fdf79010ea23d753d56d/build.rs#L7) to the code:

```Rust
    if version_check::is_max_version("1.38.0").unwrap_or(false)
        || !version_check::Channel::read().unwrap().is_stable() // Fails in the unwrap here
    {
        println!("cargo:rustc-cfg=skip_ui_tests");
    }
```

This is inside a `||` but I think it's due to Channel simply being `None`, which might be due to the try build being a local build instead of a "proper" nightly.

## 1 UTC offset issue

Example [log](https://crater-reports.s3.amazonaws.com/pr-103293/try%2351ad93974f526eaf4ede64932870490e6c96e4fd/reg/rttrust-0.1.3/log.txt):

```
[INFO] [stderr]   thread 'main' panicked at 'Could not determine the UTC offset on this system. Possible causes are that the time crate does not implement "local_offset_at" on your system, or that you are running in a multi-threaded environment and the time crate is returning "None" from "local_offset_at" to avoid unsafe behaviour. See the time crate's documentation for more information. (https://time-rs.github.io/internal-api/time/index.html#feature-flags): IndeterminateOffset', /opt/rustwide/cargo-home/registry/src/github.com-1ecc6299db9ec823/simple_logger-1.16.0/src/lib.rs:409:85
```

I think this might be a multithreading issue? I don't think that the PR caused this either.

--------

#TLDR

My impression is that the crater check found one valid build failure caused by this PR, which is very easy to fix and for which I offer to make a PR for. There was also a lot of noise from spurious failures or ones I couldn't reproduce.

I would love to hear from others on this but I think we can go forward with the PR?

danielhenrymantilla commented:
@est31 I confirm the lifetime issue you've observed is not a compiler bug you'd have introduced; it's just a matter of type inference hindering the heuristics that allow for "higher-order-signature promotion" of closures, which is necessary to be able to feed two references with _disjoint_ lifetimes (or just _different_, in the invariant case) to the same closure.

  - (You can check the documentation of <https://docs.rs/higher-order-closure> for more info about hit)

joshtriplett commented:
> The other behaviour that from my POV should be improved is the drop order of let temporaries regarding the else block. You can either choose consistency with `if let` and `match` here, or on the other side, with `if` and `let else` (edit: see #103248 as the issue). Personally I'd prefer the latter two as it frees up resources earlier, but feel free to disagree (it might be surprising to people that adding/removing a bool conditional changes what the still-existing `let` bindings are doing regarding the `else` drop order).

I have the same concern as your parenthetical. I think it's important to be consistent with `if let` here.

joshtriplett commented:
I think this is ready to consider for merging:

@rfcbot merge

@rfcbot concern consistency-with-if-let
(Per my comment above.)

joshtriplett commented:
Gah.

@rfcbot cancel

joshtriplett commented:
@rfcbot merge

@rfcbot concern consistency-with-if-let
Per https://github.com/rust-lang/rust/pull/103293#issuecomment-1293408574

est31 commented:
> I have the same concern as your parenthetical. I think it's important to be consistent with if let here.

fair, I rest my case then: I have closed #103248. The if let chains implementation on master is already consistent with if let regarding else drop ordering, so there needs to be no change towards consistency. All that was needed was a statement from the lang team (or members) that this consistency is wanted, which you just gave. This particular PR never attempted to change anything about drop order regarding else, it was only about && and || drop order. So think your concern can be removed?

joshtriplett commented:
@rfcbot resolved consistency-with-if-let

joshtriplett commented:
@est31 Thanks!

RalfJung commented:
> For the other failures, most of them are either in proc macros or build.rs failures. I need to investigate more closely, but a lot of crate failures are due to libc's build.rs failing:

FWIW that's just https://github.com/rust-lang/crater/issues/663

nikomatsakis commented:
@rfcbot reviewed

The crater run doesn't give evidence that we aren't changing runtime semantics, but it seems equally likely to me that this change fixes latent bugs as causes them.

est31 commented:
Good point, I've just checked and crater does not run any tests, even though it calls the status test-pass. It seems to just *compile* the tests? IDK. There is some minor "run time" of proc macros but that only applies to a fraction of Rust code crater builds (the proc macro crates and their dependencies).

RalfJung commented:
Crater can be run in check-only mode or told to also run tests.

est31 commented:
@RalfJung Oh so you can choose to do a test crater run vs a check one. I didn't know that this was possible so when I asked for this PR to have a crater run I didn't request it. Feel free to issue another crater run that runs the tests too if you feel that it is useful. Ultimately, impact will also be seen by the beta crater runs.

RalfJung commented:
I don't know if it will be useful, I have not been following closely. I just noted that `craterbot check` indeed explicitly asks crater to do a check-only build and nothing else, so that is what it did. :)

@craterbot run mode=build-and-test

RalfJung commented:
@craterbot run mode=build-and-test crates=https://crater-reports.s3.amazonaws.com/pr-103293-1/retry-regressed-list.txt

est31 commented:
Alright there is actual non-false-positive test failures. E.g. I could consistently reproduce the regression of the `vec3::tests::random_unit_vector_test` test in [`raytracing_rust`](https://github.com/muramasa8191/raytracing_rust). The only place in the code base that uses && is:

```Rust
pub struct Vec3 {
    x: f64,
    y: f64,
    z: f64,
}
// ...
impl Vec3 {
// ...
    pub fn near_zero(&self) -> bool {
        let s: f64 = 1e-8;
        self.x.abs() < s && self.y.abs() < s && self.z < s
    }
// ...
}
```

Yes, the temporaries are floats, and yes, that && chain is the culprit: if I add a `true &&` then I also break the test on the unmodified toolchain. The weird thing however is that `near_zero` is not being called by the `random_unit_vector_test` test, neither directly nor indirectly.

I would also say that the test is highly brittle: it first normalizes a vector, and then asserts that the squared L2 norm is smaller than 1... that's not what normalization is supposed to do, which is to make the length precisely one. It might not make it precisely 1 100% of the time due to float weirdnesses, and I guess code previously relied on that being the case, but something due to the change in this unrelated piece of code has caused the change. If the `<` in the assertion is changed to `<=` then it works again. You'd probably ideally want to have an epsilon.

I also tried reproducing the regressions of [algae](https://github.com/Osrepnay/algae), [ctw](https://github.com/davidrusu/ctw), [leetcode](https://github.com/lavrxxx/leetcode) and [bookworm](https://github.com/csauve/bookworm) but couldn't reproduce them.

Aaron1011 commented:
@est31: The `random_unit_vector_test` randomly fails for me on both stable and nightly, and adding `true &&` doesn't change this for me. Is this PR causing it to fail 100% of the time for you?

est31 commented:
@Aaron1011 I re-ran the test a bunch of times using beta. It seems to sometimes fail, sometimes work. It worked 4 times out of 20 times I tested it. So I'd say it's flaky. In the repeat crater run it now shows up as test-fail on master too, so I think we can safely discard it.

est31 commented:
Thanks @RalfJung for re-running the crater run, that helped with narrowing down the list. It seems there are now only 9 crates to look at (the test-pass -> build-fail ones would have been caught in the earlier run too). I will do a try to reproduce them locally once I find the opportunity.

est31 commented:
I've looked at the 9 crates that went from test-pass to test-fail, and most of them had flaky tests, aka, sometimes the test works, sometimes it doesn't, even if you use the nightly toolchain. That's what I used for local testing. There were two crates where I couldn't reproduce the failures: `big-o-test` was doing performance measurement. I think this is something that will likely break in a cloud CI environment. As for `testdir`, I couldn't reproduce the failures either but I have no idea what was going wrong here. It doesn't impl `Drop` anywhere however, and it also doesn't contain `||` outside of closures. It contains `&&` in [one place](https://hg.sr.ht/~flub/testdir/browse/src/numbered_dir.rs?rev=35156fc48f6769966e240461d367c0038d0dbef9#L159) when comparing `u16`s. I therefore don't think the test failure was caused by this PR.

`loam` was different in that the first test run after clearing `target` always worked, then it changes to always not working, it seems it writes some files inside the `target` dir or something like that.

| project | verdict |
| - | - |
| [codesections/dlarm](https://github.com/codesections/dlarm) | flaky tests |
| [lythesia/ruduo](https://github.com/lythesia/ruduo) | flaky tests |
| [muratamuu/datamanager](https://github.com/muratamuu/datamanager) | flaky tests |
| [yanfenglee/paste](https://github.com/yanfenglee/paste) | flaky tests |
| big-o-test 0.2.4 | couldn't reproduce failures |
| ijson 0.1.3 | flaky tests |
| kcfg 0.1.2 | flaky tests |
| loam 0.1.0 | first test run always works, then changes |
| testdir 0.5.1 | couldn't reproduce failures |

All in all, it seems that the crater run found no test runtime regression caused by this PR :tada: .

WaffleLapkin commented:
@stepancheg note that `&&` short-circuits, so it's more similar to
```rust
if F(1).g() {
    if F(2).g() {
        if F(3).g() {
            if F(4).g() {
                return true
            }
        }
    }
}
false
```

Which also drops `1`, `2`, `3`, `4`.

stepancheg commented:
@WaffleLapkin I expected your example to drop 4,3,2,1. There's something I don't understand about it. Thank you for the example!

est31 commented:
@stepancheg

> Is it correct that drop order is 1,2,3,4, not 4,3,2,1?

In theory, yes, we could drop in reverse order and not in evaluation order. And indeed, most constructs, e.g. even non-short-circuiting & and | do it that way. However, those constructs all *return* data, meaning they create something that later code could rely on. E.g. a let statement's binding could be used by the code following the let. So it's not safe to drop the temporaries of the let statement until the let's binding goes out of scope, and you have to drop let statement's temporaries in reverse order. For &, it can return any type, as it is extensible via traits, including a type that might contain a lifetime and e.g. a reference to earlier data. So here too we can't drop in reading order. But with &&, it's not extensible via traits, so the return type *must* be `bool`, which obviously can't contain any references to the temporaries, so we can drop its temporaries just as we drop the temporaries of a function that we call.

That's my justification: it's not needed to drop in reverse order, and the earlier we drop something, the better, as it helps save memory usage (and I think it's also friendlier to some optimizer passes to have less stuff around). There is also the other reason that this change is smaller than making && drop in reverse order. @WaffleLapkin 's argument is also valid, and might be closer to what the original design philosophy was according to the comments in the code region where this PR is making changes (added by 419ac4a1b899ba88fb360b4c71c08f3610564cd4 btw). So: the temporaries of the bodies with possible repetition like `loop` and `for` have to be dropped on the spot otherwise there is infinite stack growth. And we put conditional bodies into that as well (I'm not 100% following here why but shrug), and as && is short circuiting, it's a conditional body too. With *that* logic alone, you can say that the very first member of the && chain is not a conditional body and thus the current behaviour before this PR is correct. But on the other hand, the temporaries created by the if's expression also get dropped quickly, even though that entire expression isn't a conditional body either. Same goes for statements consisting of expressions. So this [prints](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bf19c521ad7b4a6a5d6c3811b3cea5dd) 1,2,3,4:

```Rust
fn main() {
    if F(1).g() {}
    println!("2");
    F(3);
    F(4);
}
```

stepancheg commented:
@est31 

> But with &&, it's not extensible via traits, so the return type must be bool, which obviously can't contain any references to the temporaries

I'm also thinking, one day rust may want to extend `&&` syntax to allow references to bindings in previous expressions. Like this:

```
if let Some(s) = s_opt && s.is_upper() && let Some(first) = s.chars().next() && first != '_' {
  ..
}
```

and to make it work, we would need reverse drop order.

> That's my justification: it's not needed to drop in reverse order, and the earlier we drop something, the better, as it helps save memory usage

Got it.

est31 commented:
r? @RalfJung

RalfJung commented:
This is not code I am familiar with, I cannot review this.
r? compiler

nagisa commented:
@nikomatsakis [said](https://github.com/rust-lang/rust/pull/103293#issuecomment-1307731377):

> rfcbot reviewed
> 
> The crater run doesn't give evidence that we aren't changing runtime semantics, but it seems equally likely to me that this change fixes latent bugs as causes them.

For some reason your checkbox is unchecked, and seems like the rfcbot never actually checked it. A bug in rfcbot?

nagisa reviewed (commented):
This PR is changing the behaviour of mixed `&&` and `||` chains as well. Please adjust a test to exercise those scenarios as well.

Especially keep in mind the fact that operator precedence for `&&` and `||` is different, so `a && b || c && d` has a different evaluation order compared to e.g. `a && (b || c) && d`. I believe with this change the difference cannot be observed either way, though.

nagisa commented on compiler/rustc_hir_analysis/src/check/region.rs:
```diff
@@ -241,18 +241,35 @@ fn resolve_expr<'tcx>(visitor: &mut RegionResolutionVisitor<'tcx>, expr: &'tcx h
                 // For shortcircuiting operators, mark the RHS as a terminating
                 // scope since it only executes conditionally.
 
+                // If the LHS is not another binop itself of the same kind as ours,
```
FWIW there are a couple of implied contexts that aren’t necessarily immediately obvious. For example, why is it that only the LHS matters, and not RHS (because LAnd/LOr binary ops are left-associative.) The double negative makes this somewhat difficult to understand as well.

Something like the following might be easier to grok, perhaps?

> We want to make sure that operands in expressions like `a && b && c` and `a || b || c` are dropped in order. These operators are left-associative, so we don’t want to terminate the LHS expression, if LHS is the same binary expression as the current expression.

At least that’s what I understand this code to be doing.

est31 commented on src/test/ui/drop/drop_order.rs:
```diff
@@ -118,17 +118,71 @@ impl DropOrderCollector {
+        let _ = self.option_loud_drop(7).is_none() // 1
+            || self.option_loud_drop(8).is_none() // 2
+            || self.option_loud_drop(9).is_none(); // 3
+        self.print(10); // 4
```
> Is it testing that the operands to the expression above are dropped before the next statement?

Yes. I can add a comment that explains this, but note that printing in the body is done with tests for other constructs in the file too.

est31 committed:
Improve comments

est31 committed:
Also avoid creating a terminating scope in mixed chains

This avoids creation of a terminating scope in
chains that contain both && and ||, because
also there we know that a terminating scope is
not neccessary: all the chain members are already
in such terminating scopes.

Also add a mixed && / || test.

nagisa commented:
@bors r+

bors commented:
:pushpin: Commit a59a2d3f6a80001b0610f03cfc7a6452b63f8935 has been approved by `nagisa`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nagisa a59a2d3f6a80001b0610f03cfc7a6452b63f8935 -->
<!-- homu: {"type":"Approved","sha":"a59a2d3f6a80001b0610f03cfc7a6452b63f8935","approver":"nagisa","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit a59a2d3f6a80001b0610f03cfc7a6452b63f8935 with merge 19c250aa1231f1e1b0953829e999e4862961207d...
<!-- homu: {"type":"BuildStarted","head_sha":"a59a2d3f6a80001b0610f03cfc7a6452b63f8935","merge_sha":"19c250aa1231f1e1b0953829e999e4862961207d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3613593171/jobs/6090337638)
Approved by: nagisa
Pushing 19c250aa1231f1e1b0953829e999e4862961207d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3613593171/jobs/6090337638"},"merge_sha":"19c250aa1231f1e1b0953829e999e4862961207d"} -->

rust-timer commented:
Finished benchmarking commit (19c250aa1231f1e1b0953829e999e4862961207d): [comparison URL](https://perf.rust-lang.org/compare.html?start=344889e963742e87181d3c023e2a1ea7d95f9468&end=19c250aa1231f1e1b0953829e999e4862961207d&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=344889e963742e87181d3c023e2a1ea7d95f9468&end=19c250aa1231f1e1b0953829e999e4862961207d&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=344889e963742e87181d3c023e2a1ea7d95f9468&end=19c250aa1231f1e1b0953829e999e4862961207d&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=344889e963742e87181d3c023e2a1ea7d95f9468&end=19c250aa1231f1e1b0953829e999e4862961207d&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
